valheim-plus-updater

```diff
@@ -0,0 +1,88 @@
+    local api_response=$(curl -sL https://api.github.com/repos/nxPublic/ValheimPlus/releases/latest | jq -r ".assets[] | select(.name == \"$vp_zipfile\")")
+    local download_url=$(jq -r  '.browser_download_url' <<< "${api_response}" )
+    local remote_updated_at=$(jq -r  '.updated_at' <<< "${api_response}" )
+    local remote_size=$(jq -r  '.size' <<< "${api_response}")
+
+    if [ -f "$vp_zipfile" -a -f "$vp_zipfile.updated_at" ]; then
+        local_updated_at=$(< "$vp_zipfile.updated_at")
+        local_size=$(stat --format="%s" "$vp_zipfile")
+        if [ "$local_updated_at" = "$remote_updated_at" -a "$local_size" = "$remote_size" ]; then
+            echo "Local ValheimPlus archive is identical to remote archive - no update required"
+        else
+            echo "Local ValheimPlus archive with size $local_size and update date $local_updated_at differs from remote size $remote_size and date $remote_updated_at - updating"
+            prepare_valheim_plus "$download_url" "$remote_updated_at"
+        fi
+    else
+        echo "Fresh ValheimPlus install"
+        prepare_valheim_plus "$download_url" "$remote_updated_at"
+    fi
+    if [ -f "$vp_mergefile" ]; then
+        echo "Valheim dedicated server or ValheimPlus got updated - merging installation files"
+        rm -f "$vp_mergefile"
+        merge_valheim_plus
+    fi
+}
+
+prepare_valheim_plus() {
+    local download_url="$1"
+    local updated_at="$2"
+    download_valheim_plus "$download_url" $updated_at \
+        && extract_valheim_plus \
+        && touch "$vp_mergefile"
+}
+
+merge_valheim_plus() {
+    rm -rf "$vp_install_path.tmp" "$vp_install_path.old"
+    mkdir -p "$vp_install_path.tmp"
+    rsync -a --itemize-changes --exclude server_exit.drp --exclude steamapps "$valheim_download_path/" "$vp_install_path.tmp"
+    rsync -a --itemize-changes "$vp_download_path/extracted/" "$vp_install_path.tmp"
+    if [ ! -d "/config/valheimplus" ]; then
+        cp -a "$vp_install_path.tmp/BepInEx/config" "/config/valheimplus"
+    fi
+    rm -rf "$vp_install_path.tmp/BepInEx/config"
+    chmod +x "$vp_install_path.tmp/start_game_bepinex.sh"
+    ln -s /config/valheimplus "$vp_install_path.tmp/BepInEx/config"
+    if [ -d "$vp_install_path" ]; then
+        mv -f "$vp_install_path" "$vp_install_path.old"
+    fi
+    mv "$vp_install_path.tmp" "$vp_install_path"
+    if [ ! -f "$valheim_restartfile" ]; then
+        echo restart > "$valheim_restartfile"
+    fi
+}
+
+download_valheim_plus() {
+    local download_url="$1"
+    local updated_at="$2"
+    local download_path="$vp_download_path/$vp_zipfile"
+    echo "Downloading $download_url to $download_path"
+    curl -L -o "$download_path" "$download_url" \
+        && echo "$updated_at" > "$download_path.updated_at"
+}
+
+extract_valheim_plus() {
+    cd "$vp_download_path"
+    echo "Extracting downloaded ValheimPlus ZIP archive"
+    rm -rf extracted
+    mkdir -p extracted
+    unzip -d extracted/ $vp_zipfile
```

Comment:
You should consider only extracting the folders `BepInEx`, `unzipped_corlib`, and `doorstop_libs` as the rest of the files are not necessary

Reply:
Hmm in that case I would likely want to extract everything but then only sync the relevant folders in the merge process. Sorta like we do with the valheim server Steam download where some stuff gets excluded. At the same time, if they introduce new directories we don't have to change the code and the extra files don't seem to hurt. So I'm torn on this one.

Reply:
Is there anything you would want to explicitly exclude? Maybe the .sh scripts?

Reply:
In the same vein, they have renamed the start_server_bepinex.sh script at least once based on the documentation. Also incorporating all of the exports from that shell script into the valheim-server code means that you now have to maintain and mirror any changes that are pushed to that script.

Wouldn't it just be safer to keep the entire mod, then just find and remove the line that calls `./valheim_server`, then the run_server method could just run the mod's shell script to export the needed values, and launch the server with the config passed into the docker?